[MouseHighlighter] bringing window to front if pushed lower in Z order

donlaci opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
[MouseHighlighter] Fixing issue MousHighlighter not visible when clicking on 'pinned to top' window.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [X] **Closes:** #23456 
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
The faulty scenario is that there is a window pinned to top and not focused, MouseHighlighter is on. 
When the user clicks to the window, the window gets focused and is moved to the top in Z order causing our window getting hidden by that window. To solve that we have to bring our MouseHighlighter window to top position again, after the pinned window processes the mouse event. To do that a timer is started on mouse down event (when we draw the highlighter circles), the timer period is set to 10 ms. After the timer rings we bring our window to the front again. As we experience that it takes 0-30 ms that the pinned window hides our window, we await 5 timer ticks (50 ms together) and then we stop the timer. 

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
tested locally

donlaci committed:
[MouseHighlighter] modifying window to be 'topmost' to be visible when clicking on 'pinned to top' windows.

stefansjfw reviewed (commented):
Looks like this affects WinAppSDK apps. Tested with WinUI galery, PT Settings, PowerRename, etc.. This change does not fix the mentioned behavior

donlaci committed:
Fixing window hidden when clicking on a pinned to top window.

jaimecbernardo reviewed (changes requested):
From what I saw, this is just firing lots of times in a loop.
I added some debugs to check behavior, and as soon as I pressed Windows+Shift+H, it just started looping timers:
![image](https://github.com/microsoft/PowerToys/assets/26118718/5c3b9da5-aaf4-438d-b3d1-69775e870070)
That's leaking a lot, can't go on like this.
Also seems to be leaking timers, since it seems to start 3 at once?

donlaci committed:
Moving the code which starts the bring-to-front-timer to the mouse click events.
Also adding modifications to await 5 times 10 ms after the event to have surely the desired effect as in some cases the window repositioning takes longer than 10 ms.

davidegiacometti commented:
If I remember correctly mouse utilities are using hotkeys instead of LLKH. This may be related https://github.com/microsoft/PowerToys/issues/11358

donlaci committed:
Minor changes

donlaci committed:
Add safety code to ensure timer is never started paralely. Also minor format changes.

jaimecbernardo commented on src/modules/MouseUtils/MouseHighlighter/MouseHighlighter.cpp:
```diff
@@ -330,6 +342,23 @@ LRESULT CALLBACK Highlighter::WndProc(HWND hWnd, UINT message, WPARAM wParam, LP
+        {
+        case BRING_TO_FRONT_TIMER_ID:
+            static int fireCount = 0;
+            if (fireCount++ >= 4)
```
Why are we firing the timer 4 times before de-registering?
Also the description is outdated regarding what the PR is doing, right?

donlaci replied:
I've updated the description, thanks for the hint.
My experience is that it takes 0-30ms that our window gets hidden by the pinned window. If we use a simple timer with one period of 50ms, there is a little vibration on the UI, as in most of the cases our circle gets hidden in a few miliseconds. One timer with a short period is not good enough, in 5% of the cases the pinned window will hide our window. To have a flickeringless UI, we bring our window to top position 5 times (with 10 ms delay between them).

jaimecbernardo replied:
Let's add some comments to explain it, then, please 😉

donlaci committed:
Add comments

donlaci committed:
spell checker corrections

jaimecbernardo reviewed (changes requested):
Gave it a try with the most recent changes and it doesn't seem to be working for me in all cases. For example, I set a notepad windows to be always on top and try to press the title bar. The highlight shows for a little bit and then goes a way. Sometimes it even blinks for a bit.

donlaci committed:
Merge remote-tracking branch 'origin/main' into MouseHighlighter-AlwaysTop

jaimecbernardo reviewed (approved):
Code LGTM! And behavior is good enough. I don't think we can work around Windows being aggressive.
Thank you!
